---
layout: page
title: NanoMod
navigation: 5
---

# NanoMod
This module allows to predict the loci with RNA modifications. Data produced by NanoPreprocess is required as input. Moreover, reads must be aligned to the transcriptome.

## Workflow

<img src="https://biocorecrg.github.io/master_of_pores_dev/dag_mod_v3.png" width="600" align="middle">

| Process name  | Software | Description |
| ------------- | ------------- | ------------- |
| **IndexReferenceForEpinano** | Epinano | Index the reference file to run Epinano. |
| **CallVariantsForEpinano** | Epinano | Perform variant calling with Samtools. |
| **calcVarFrequenciesForEpinano** | Epinano | Frequencies calculation for each Epinano feature (mismatch, insertion, deletion) and quality analysis. |
| **indexing_with_nanopolish** | Nanopolish | Indexing using the sequencing summary file, fast5 and fastq files. |
| **eventalign_with_nanopolish** | Nanopolish | Resquiggle - align signal-level events to a reference genome with Nanopolish. |
| **cat_collapsed_nanopolish** | Nanopolish | Merge all results and extract the median current value per position. |
| **calcNanoCompore** | Nanocompore | Execution of Nanocompore algorithm. |
| **resquiggling_with_tombo** | Tombo | Resquiggle - align signal-level events to a reference genome with Tombo. |
| **getModificationsWithTombo** | Tombo | RNA modification detection with Tombo (`-level_sample_compare`). |

## Input Parameters

| Parameter name  | Description [Default/Expected format] |
| ------------- | --------------|
|**input_path**| Path to the directory that contains all results generated by Nanopreprocess for each sample. **[/Path/input_folder]** |
|**comparison.tsv**| Path to the tab separated text file containing the samples to compare: see example in following section. |
|**reference**| Transcriptome in fasta format. **[Reference_transcriptome.fa]**|
|**output**| Output folder name. **[/Path/to_output_folder]**|
|**coverage**| Coverage threshold for RNA modification detection. **[5]**|
|**tombo_opt**| Command line options for Tombo. Check available options in Tombo's repository - level sample compare function. **[--num-bases 5]**|
|**epinano_opt**| Command line options for Epinano. Check available options in Epinano's repository.|
|**nanopolish_opt**| Command line options for Nanopolish. Check available options in Nanopolish's repository.|
|**nanocompore_opt**| Command line options for Nanocompore. Check available options in Nanocompore's repository. **[--sequence_context 4 --downsample_high_coverage 10000]**|
|**email**| Users email for receving the final report when the pipeline is finished. **[user_email]**|

## Running the pipeline
In order to run NanoMod, it is required that Nanopreprocess results are stored in a directory with the following structure:
```
Nanopreprocess_output  
│
└───wt_1
│   └───fast5_files
│   └───fastq_files
│   └───alignment
│   └───QC_files
│   
└───wt_2
│   └───fast5_files
│   └───fastq_files
│   └───alignment
│   └───QC_files
|
└───ko_1
│   └───fast5_files
│   └───fastq_files
│   └───alignment
│   └───QC_files
|
└───ko_4
│   └───fast5_files
│   └───fastq_files
│   └───alignment
│   └───QC_files
```

Then, user should fill in all input file based on the directory shown above:
* In `params.config` file: `**input_path** = /Path_to_directory/Nanopreprocess_output`
* `comparison.tsv` file - please see example below:
```bash
wt_1 ko_1
wt_2 ko_2
...
```
<br/>

Once the setting up is done, please use the following commands to run NanoMod:
```bash
nextflow run nanomod.nf -with-singularity > log.txt
```
<br/>

* Run the pipeline in the background:
```bash
nextflow run nanomod.nf -with-singularity -bg > log.txt
```
<br/>

* Run the pipeline while changing **params.config** file:
```bash
nextflow run nanomod.nf -with-singularity -bg --output test2 > log.txt
```
<br/>

* Specify a directory for the working directory (temporary files location):
```bash
nextflow run nanopmod.nf -with-singularity -bg -w /path/working_directory > log.txt
```

## Results
NanoMod creates four folders which contain all results generated by each one of the softwares (Epinano, Nanopolish, Tombo and Nanocompore). 

1. **Epinano**: two files per sample. One containing data at position level and the other, at 5-mer level. Different features frequencies as well as quality data are included in the results. See example below: 

```bash
#Ref,pos,base,cov,q_mean,q_median,q_std,mis,ins,del
gene_A,2515,C,45497.0,5.36995,4.00000,3.97797,0.0822032221904741,0.18715519704595907,0.2058377475437941
gene_A,2516,A,45504.0,5.38207,4.00000,4.71619,0.17128164556962025,0.20497099156118143,0.07733386075949367
gene_A,2517,C,45529.0,6.92130,5.00000,5.04250,0.06165301236574491,0.1505633771881658,0.13540820136616222
gene_A,2518,A,45545.0,6.49821,5.00000,5.47485,0.10802503018992206,0.10855198155670216,0.2082775277198375
gene_A,2519,T,45557.0,6.51247,5.00000,4.81853,0.09386043857145993,0.14792457800118533,0.2033057488421099
gene_A,2520,C,45609.0,10.72107,9.00000,7.03936,0.03990440483237957,0.1465938740160933,0.020105680896314326
gene_A,2521,C,45613.0,13.44950,13.00000,7.70959,0.041479402801832814,0.07280819064740315,0.1242189726613027
gene_A,2522,T,45621.0,13.79766,14.00000,7.79424,0.07599570373293,0.0977181561123167,0.050897612941408564
...
```
<br/>

2. **Nanopolish**: two files per sample - raw eventalign output (gzipped) and another with the median raw current per position and transcript. See example below:
```bash
contig	position	reference_kmer	read_name	median	coverage
gene_A	0	AAATT	1	113.35	433
gene_A	1	AATTG	1	97.24	506
gene_A	2	ATTGA	1	70.35	2034
gene_A	3	TTGAA	1	102.03	416
gene_A	4	TGAAG	1	115.315	422
gene_A	5	GAAGA	1	104.25	471
gene_A	6	AAGAG	1	122.785	230
gene_A	7	AGAGT	1	132.3	314
...
```
<br/>

3. **Tombo**: one file per comparison. It reports the p-value per position, the sum of p-values per 5-mer and coverage in both WT and KO. See example below:
```bash
"Ref_Position"	"Chr"	"Position"	"Tombo_SiteScore"	"Coverage_Sample"	"Coverage_IVT"	"Tombo_KmerScore"
"gene_A_3"	"gene_A"	"3"	"0.0000"	"92"	"87"	NA
"gene_A_4"	"gene_A"	"4"	"0.0000"	"92"	"87"	NA
"gene_A_5"	"gene_A"	"5"	"0.0000"	"92"	"87"	0
"gene_A_6"	"gene_A"	"6"	"0.0000"	"93"	"88"	0.0014
"gene_A_7"	"gene_A"	"7"	"0.0000"	"95"	"89"	0.0027
"gene_A_8"	"gene_A"	"8"	"0.0014"	"95"	"89"	0.004
"gene_A_9"	"gene_A"	"9"	"0.0013"	"96"	"90"	0.004
"gene_A_10"	"gene_A"	"10"	"0.0013"	"136"	"130"	0.004
...
```
<br/>

4. **Nanocompore**: one file per comparison. Default output from Nanocompore (see Nanocompore's repository for a more detailed explanation). 
add
